stream-chat-android-offline/src/main/java/io/getstream/chat/android/offline/event/handler/internal/batch/SocketEventCollector.kt

```diff
@@ -0,0 +1,161 @@
+/*
+ * Copyright (c) 2014-2022 Stream.io Inc. All rights reserved.
+ *
+ * Licensed under the Stream License;
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    https://github.com/GetStream/stream-chat-android/blob/main/LICENSE
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.getstream.chat.android.offline.event.handler.internal.batch
+
+import io.getstream.chat.android.client.events.ChatEvent
+import io.getstream.chat.android.client.events.UserStartWatchingEvent
+import io.getstream.chat.android.client.events.UserStopWatchingEvent
+import io.getstream.logging.StreamLog
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+
+/**
+ * Collects socket events to produce a single [BatchEvent].
+ */
+internal class SocketEventCollector(
+    private val scope: CoroutineScope,
+    private val fireEvent: suspend (BatchEvent) -> Unit,
+) {
+    private val logger = StreamLog.getLogger("Chat:EventCollector")
+    private val mutex = Mutex()
+    private val postponed = Postponed()
+    private val timeoutJob = TimeoutJob()
+
+    internal suspend fun add(event: ChatEvent): Boolean {
+        if (event is UserStartWatchingEvent || event is UserStopWatchingEvent) {
+            logger.d { "[add] event.type: ${event.type}(${event.hashCode()})" }
+            mutex.withLock {
+                timeoutJob.cancel()
+                return postponed.add(event).also {
+                    when {
+                        postponed.size >= ITEM_COUNT_LIMIT -> onItemCountLimit()
+                        postponed.collectionTime() >= TIME_LIMIT -> onTimeLimit()
+                        else -> scheduleTimeout()
```

Comment:
If I understand this correctly, we are always waiting 300ms to batch and fire events. In the case of a single event or a sparse array of events, this will slow performance (it doesn't trigger events right away). Couldn't we keep this logic, but handle this with some kind of backpressure instead of always batching events by time?